bindings/rust/evmc-declare/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+fn build_create_fn(names: &VMNameSet) -> proc_macro2::TokenStream {
+    let type_ident = names.get_type_as_ident();
+    let fn_ident = names.get_lowercase_as_ident_prepend("evmc_create_");
+
+    let static_version_ident = names.get_caps_as_ident_append("_VERSION");
+    let static_name_ident = names.get_caps_as_ident_append("_NAME");
+
+    quote! {
+        #[no_mangle]
+        extern "C" fn #fn_ident() -> *const ::evmc_sys::evmc_instance {
+            let new_instance = ::evmc_sys::evmc_instance {
+                abi_version: ::evmc_sys::EVMC_ABI_VERSION as i32,
+                destroy: Some(__evmc_destroy),
+                execute: Some(__evmc_execute),
+                get_capabilities: Some(__evmc_get_capabilities),
+                set_option: None,
+                set_tracer: None,
+                name: ::std::ffi::CString::new(#static_name_ident).expect("Failed to build VM name string").into_raw() as *const i8,
+                version: ::std::ffi::CString::new(#static_version_ident).expect("Failed to build VM version string").into_raw() as *const i8,
+            };
+
+            unsafe {
+                ::evmc_vm::EvmcContainer::into_ffi_pointer(Box::new(::evmc_vm::EvmcContainer::<#type_ident>::new(new_instance)))
+            }
+        }
+    }
+}
+
+/// Builds a callback to dispose of the VM instance
+fn build_destroy_fn(names: &VMNameSet) -> proc_macro2::TokenStream {
+    let type_ident = names.get_type_as_ident();
+
+    quote! {
+        extern "C" fn __evmc_destroy(instance: *mut ::evmc_sys::evmc_instance) {
+            unsafe {
+                ::evmc_vm::EvmcContainer::<#type_ident>::from_ffi_pointer(instance);
+            }
+        }
+    }
+}
+
+fn build_execute_fn(names: &VMNameSet) -> proc_macro2::TokenStream {
+    let type_name_ident = names.get_type_as_ident();
+
+    quote! {
+        extern "C" fn __evmc_execute(
+            instance: *mut ::evmc_sys::evmc_instance,
+            context: *mut ::evmc_sys::evmc_context,
+            rev: ::evmc_sys::evmc_revision,
+            msg: *const ::evmc_sys::evmc_message,
+            code: *const u8,
+            code_size: usize
+        ) -> ::evmc_sys::evmc_result
+        {
+            assert!(!msg.is_null());
+            assert!(!context.is_null());
+            assert!(!instance.is_null());
+            assert!(!code.is_null());
+
+            let execution_context = unsafe {
+                ::evmc_vm::ExecutionContext::new(
+                    msg.as_ref().expect("EVMC message is null"),
+                    context.as_mut().expect("EVMC context is null")
+                )
+            };
+
+            let code_ref: &[u8] = unsafe {
+                ::std::slice::from_raw_parts(code, code_size)
+            };
+
+            let container = unsafe {
+                ::evmc_vm::EvmcContainer::<#type_name_ident>::from_ffi_pointer(instance)
+            };
+
+            let result = container.execute(code_ref, &execution_context);
```

Comment:
Should wrap this in panic unwinding and return `EVMC_INTERNAL_ERROR`:
```
panic::catch_unwind(|| {
    container.execute(..)
});
```

Reply:
Well actually we should do that for the entire `execute` function except the `assert!`s on top.

Reply:
And potentially for panics caused but this macro should be ensured that result in abort and not unwinding.

Reply:
Created issue #310 to track this.

Reply:
I'd be cautious about trying to introduce unwinding infrastructure this early. Maybe in a separate PR.